sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/HybridSearchDocumentQueryExecutionContext.java

```diff
@@ -0,0 +1,500 @@
+                return results;
+            }).flatMapMany(Flux::fromIterable);
+    }
+
+    private static Mono<List<List<Integer>>> computeRanks(Mono<List<List<ScoreTuple>>> componentScoresList) {
+        return componentScoresList.flatMap(componentScores -> {
+            // initialize ranks as an N-D list with zeros
+            List<List<Integer>> ranksInternal = new ArrayList<>();
+            for (int componentIndex = 0; componentIndex < componentScores.size(); componentIndex++) {
+                List<Integer> row = new ArrayList<>();
+                for (int index = 0; index < componentScores.get(0).size(); index++) {
+                    row.add(0);
+                }
+                ranksInternal.add(row);
+            }
+
+            for (int componentIndex = 0; componentIndex < componentScores.size(); componentIndex++) {
+                int rank = 1; // ranks are 1 based
+                for (int index = 0; index < componentScores.get(componentIndex).size(); index++) {
+                    // Identical scores should have the same rank
+                    if ((index > 0) && (componentScores.get(componentIndex).get(index).getScore() < componentScores.get(componentIndex).get(index - 1).getScore())) {
+                        rank += 1;
+                    }
+                    int rankIndex = componentScores.get(componentIndex).get(index).getIndex();
+                    ranksInternal.get(componentIndex).set(rankIndex, rank);
+                }
+            }
+            return Mono.just(ranksInternal);
+        });
+    }
+
+    private static Mono<List<List<ScoreTuple>>> retrieveComponentScores(Mono<List<HybridSearchQueryResult<Document>>> coalescedAndSortedResults) {
+        return coalescedAndSortedResults.flatMap(results -> {
+            List<List<ScoreTuple>> componentScoresInternal = new ArrayList<>();
+            for (int i = 0; i < results.get(0).getComponentScores().size(); i++) {
+                componentScoresInternal.add(new ArrayList<>());
+            }
+            for (int i = 0; i < results.size(); i++) {
+                List<Double> componentScores = results.get(i).getComponentScores();
+                for (int j = 0; j < componentScores.size(); j++) {
+                    ScoreTuple scoreTuple = new ScoreTuple(componentScores.get(j), i);
+                    componentScoresInternal.get(j).add(scoreTuple);
+                }
+            }
+            //Sort scores in descending order
+            for (List<ScoreTuple> scoreTuples : componentScoresInternal) {
+                scoreTuples.sort(Comparator.comparing(ScoreTuple::getScore, Comparator.reverseOrder()));
+            }
+            return Mono.just(componentScoresInternal);
+        });
+    }
+
+    private Flux<Document> getComponentQueryResults(List<FeedRangeEpkImpl> targetFeedRanges, int initialPageSize, DocumentCollection collection, Flux<QueryInfo> rewrittenQueryInfos) {
+        return rewrittenQueryInfos.flatMap(queryInfo -> {
```

Comment:
Concurrency for flatMap

Reply:
Ditto as above: I am doing a flatMap on Mono, which doesn't takes in a concurrency parameter. And going over the code again we don't need a flatMap here, so I have updated the code to use a map.